### PR TITLE
uf-7203: Created API and some service layer logic.

### PR DESCRIPTION
**Pull request by mha109**

### CODE REVIEW

File: src/main/java/se/sundsvall/billingdatacollector/api/CollectorResource.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+		@Parameter(name = "flowInstanceId", description = "flowInstanceId to trigger billing for", example = "123")
+		@NotEmpty @PathVariable("flowInstanceId") final String flowInstanceId) {
+
+		collectorService.trigger(flowInstanceId);
+
+		return ResponseEntity.accepted().build();
+	}
+
+	@Operation(
+		summary = "Trigger billing for all flowInstancesId:s between two specific dates",
```
Comment (jklingst) [inline src/main/java/se/sundsvall/billingdatacollector/api/CollectorResource.java:75]: Typo: "flowInstancesId:s"...one plural s too many

File: src/main/java/se/sundsvall/billingdatacollector/api/CollectorResource.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	content = @Content(schema = @Schema(oneOf = {Problem.class, ConstraintViolationProblem.class})))
+@ApiResponse(
+	responseCode = "500",
+	description = "Internal Server Error",
+	content = @Content(schema = @Schema(implementation = Problem.class)))
+@ApiResponse(
+	responseCode = "502",
+	description = "Bad Gateway",
+	content = @Content(schema = @Schema(implementation = Problem.class)))
+public class CollectorResource {
```
Comment (jklingst) [inline src/main/java/se/sundsvall/billingdatacollector/api/CollectorResource.java:47]: Can be package-protected as well as the constructor

File: src/main/java/se/sundsvall/billingdatacollector/api/CollectorResource.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+			@ApiResponse(
+				responseCode = "202",
+				description = "Accepted",
+				useReturnTypeSchema = true)
+		}
+	)
+	@PostMapping(path = "/trigger", produces = { APPLICATION_JSON_VALUE, APPLICATION_PROBLEM_JSON_VALUE })
+	ResponseEntity<Void> triggerBilling(
+		@RequestParam(name = "startDate") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE)
+		@Valid @Parameter(example = "2024-01-01") final LocalDate startDate,
```
Comment (jklingst) [inline src/main/java/se/sundsvall/billingdatacollector/api/CollectorResource.java:86]: `@Valid` isn't required here

File: src/main/java/se/sundsvall/billingdatacollector/api/CollectorResource.java (hunk 1 above)
Comment (mha109) [reply src/main/java/se/sundsvall/billingdatacollector/api/CollectorResource.java:75]: yep typo

File: src/main/java/se/sundsvall/billingdatacollector/api/CollectorResource.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+		validateDateRange(startDate, endDate);
+
+		collectorService.triggerBetweenDates(startDate, endDate);
+
+		return ResponseEntity.accepted().build();
+	}
+
+	//Validate that the end date is after, or equal to, the start date
+	private void validateDateRange(LocalDate startDate, LocalDate endDate) {
+		if(startDate.isAfter(endDate)) {
```
Comment (jklingst) [inline src/main/java/se/sundsvall/billingdatacollector/api/CollectorResource.java:100]: This should be verified with regard to the OpenE "API", but I suspect that we will have to support passing the same date as from and to, and therefore, this should be something like `startDate.isEqualOrAfter(endDate)` (that doesn't exist out-of-the-box...)

File: src/test/java/se/sundsvall/billingdatacollector/api/CollectorResourceTest.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+		//Arrange
+		doNothing().when(mockService).triggerBetweenDates(Mockito.any(LocalDate.class), Mockito.any(LocalDate.class));
+
+		var startDate = LocalDate.of(2023, 4, 25);
+		var endDate = LocalDate.of(2024, 4, 25);
+
+		//Act
+		webTestClient.post()
+			.uri(uriBuilder -> uriBuilder.path("/trigger")
+				.queryParam("startDate", startDate.toString())
```
Comment (jklingst) [inline src/test/java/se/sundsvall/billingdatacollector/api/CollectorResourceTest.java:62]: Maybe explicitly use the desired date format, to be absolutely sure that relying on default date formatting doesn't f*ck things up?

File: src/test/java/se/sundsvall/billingdatacollector/api/CollectorResourceTest.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+		verify(mockService).triggerBetweenDates(date, date);
+		verifyNoMoreInteractions(mockService);
+	}
+
+	@Test
+	void testTriggerBetweenFaultyDateInterval_shouldThrowException() {
+		//Arrange & Act
+		var responseBody = webTestClient.post()
+			.uri(uriBuilder -> uriBuilder.path("/trigger")
+				.queryParam("startDate", LocalDate.of(2025, 4, 25).toString())
```
Comment (jklingst) [inline src/test/java/se/sundsvall/billingdatacollector/api/CollectorResourceTest.java:99]: Sonar saying hi here, but maybe define `"startDate"` and `"endDate"` as constants?

File: src/test/java/se/sundsvall/billingdatacollector/integration/opene/OpenEIntegrationTest.java [hunk 7]
```diff
@@ -55,15 +55,15 @@ void getErrandIds(@Load("/open-e/flow-instances.xml") final String xml) {
 	@Test
 	void getErrand(@Load("/open-e/flow-instance.internal.xml") final String xml) {
 		when(mockOpenEClient.getErrand("123456")).thenReturn(xml.getBytes(UTF_8));
-		when(mockMapper.mapToBillingRecord(any())).thenReturn(BillingRecordWrapper.builder().build());
+		when(mockMapper.mapToBillingRecordWrapper(any())).thenReturn(BillingRecordWrapper.builder().build());
```
Comment (jklingst) [inline src/test/java/se/sundsvall/billingdatacollector/integration/opene/OpenEIntegrationTest.java:58]: `any(byte[].class)` ?

File: src/test/java/se/sundsvall/billingdatacollector/api/CollectorResourceTest.java (hunk 5 above)
Comment (mha109) [reply src/test/java/se/sundsvall/billingdatacollector/api/CollectorResourceTest.java:62]: Hm I think `localDate.toString()` will output on ISO-8601-format which is `yyyy-MM-dd` ?

Comment (jklingst) [reply src/test/java/se/sundsvall/billingdatacollector/api/CollectorResourceTest.java:62]: You are correct, dear sir

Comment (mha109) [reply src/test/java/se/sundsvall/billingdatacollector/api/CollectorResourceTest.java:62]: I think I can remove the `toString` though.

File: src/test/java/se/sundsvall/billingdatacollector/integration/opene/OpenEIntegrationTest.java (hunk 7 above)
Comment (mha109) [reply src/test/java/se/sundsvall/billingdatacollector/integration/opene/OpenEIntegrationTest.java:58]: Just a rename of the method but you are correct

File: src/main/java/se/sundsvall/billingdatacollector/api/CollectorResource.java (hunk 4 above)
Comment (mha109) [reply src/main/java/se/sundsvall/billingdatacollector/api/CollectorResource.java:100]: The logic is correct though, but I'll rewrite it to make it readable

**Review by jklingst — DISMISSED** [review]

**Review by jklingst — APPROVED** [review]

**Pull request merged**